### PR TITLE
Feature/demo2x

### PR DESCRIPTION
**Pull request by mihaisoloi**
the IntegrationSpec doesn't work on master either

### CODE REVIEW

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 1]
```diff
@@ -1,58 +1,67 @@
...
         }
 
-        webSocket.onmessage = (event: MessageEvent) => {
-          downstream.onNext(event.data.asInstanceOf[String])
-        }
-      }
-      catch {
-        case ex: Throwable =>
-          downstream.onError(ex)
+        BooleanCancelable()
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: This shouldn't be an empty cancelable.

Amongst the biggest changes in Monix 2.0 is the behavior on cancelation. As you have noticed, `unsafeSubscribeFn` now returns a `Cancelable`. The purpose of this `Cancelable` is to cancel a subscription, whereas in Monifu 1.x we were doing this by a trick, by calling `takeWhileNotCanceled(cancelable)`, which requires the producer to push another item into the consumer in order to notice that it has been canceled. Therefore in Monifu 1.x the cancelation was NOT immediate.

This is why in `monifu-sample` we've got a `def createChannel(webSocket: WebSocket)` that takes an already initialized `WebSocket`. Because that `WebSocket` reference we need to close whenever something happens (like a `timeout` being triggered). If you'll see below, we have to detect problems and take care of calling `closeConnection(webSocket)` ourselves. But now that we return a `Cancelable`, this `Cancelable` can take care of calling `closeConnection(webSocket)`. In other words:
1. Make this returned cancelable call the logic in `closeConnection(webSocket)` (and note that it can be a plain `Cancelable` and not a `BooleanCancelable`)
2. Whatever subscriber is being subscribed to this `Observable`, we need to wrap it into something that also calls `closeConnection` if `onError` or `onComplete` happens (we should no longer rely on `doOnCancel`, because its meaning has changed btw ;))
3. The `createChannel()` function should no longer receive that `WebSocket` reference. All logic should be encapsulated in this function. The `closeConnection` function itself should be an inner function of this function.

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 2]
```diff
@@ -1,58 +1,67 @@
...
       }
     }
 
   private def closeConnection(webSocket: WebSocket)(implicit s: Scheduler): Unit = {
     if (webSocket != null && webSocket.readyState <= 1)
-      try webSocket.close() catch { case _: Throwable => () }
+      try webSocket.close() catch {
+        case _: Throwable => ()
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: Here we shouldn't catch `Throwable`. Not sure why I was doing it. Maybe it's because this runs on top of Javascript therefore `NonFatal` isn't so useful anymore. But we shouldn't do this anyway. And errors should be logged.

In other words `catch NonFatal(ex) => s.reportError(ex)` because otherwise we are teaching people bad practices.

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 3]
```diff
@@ -1,58 +1,67 @@
...
+
+          webSocket.onclose = (event: CloseEvent) => {
+            downstream.onComplete()
+          }
+
+          webSocket.onmessage = (event: MessageEvent) => {
+            downstream.onNext(event.data.asInstanceOf[String])
+          }
+        } catch {
+          case ex: Throwable =>
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: We shouldn't catch `Throwable` here. I think we should catch `NonFatal(ex)`. On top of Javascript it doesn't matter, but we shouldn't teach people bad practices.

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 4]
```diff
@@ -63,13 +72,13 @@ final class BackPressuredWebSocketClient private (url: String)
...
     catch {
       case ex: Throwable =>
-        Observable.error(ex)
+        Observable.raiseError(ex)
     }
 
-    val source = channel.timeout(5.seconds)
-      .doOnCanceled(closeConnection(webSocket))
+    val source = channel.dropByTimespan(1.second)
+      .doOnCancel(closeConnection(webSocket))
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: 1. I do not understand why you did `dropByTimespan`. The original code was saying that if the source goes silent for `5.seconds`, then consider that to be a timeout and reconnect.
2. `doOnCancel` no longer means what you think it means. The old behavior is now in `Observable.doOnDownstreamStop` ([see api](https://goo.gl/PoCRRH)). It doesn't matter, this call is not needed anymore, because we are going to move this logic in `createChannel`.

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 5]
```diff
@@ -78,14 +87,14 @@ final class BackPressuredWebSocketClient private (url: String)
         scheduler.reportFailure(ex)
         // retry connection in a couple of secs
         self.delaySubscription(3.seconds)
-          .onSubscribe(subscriber)
+          .subscribe(subscriber)
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: Here we should call `unsafeSubscribeFn` because we know what we are doing (i.e. our stuff is protected from errors with try/catch, we do know how to apply back-pressure, etc). Plain users are the ones that are supposed to call `subscribe`, as it wraps `Subscriber` implementations into a `SafeSubscriber` that does error handling and applies back-pressure on `onComplete` and `onError`.

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 6]
```diff
@@ -78,14 +87,14 @@ final class BackPressuredWebSocketClient private (url: String)
...
       }
 
       def onComplete(): Unit = {
         closeConnection(webSocket)
         // retry connection in a couple of secs
         self.delaySubscription(3.seconds)
-          .onSubscribe(subscriber)
+          .subscribe(subscriber)
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: Like above, I think this should be calling `unsafeSubscribeFn`.

File: client/src/main/scala/client/DataConsumer.scala [hunk 7]
```diff
@@ -26,7 +28,7 @@ final class DataConsumer(interval: FiniteDuration, seed: Long, doBackPressure: B
 
     source
       .collect { case IsEvent(e) => e }
-      .onSubscribe(subscriber)
+      .subscribe(subscriber)
```
Comment (alexandru) [inline client/src/main/scala/client/DataConsumer.scala]: This should be `unsafeSubscribeFn`.

File: client/src/main/scala/client/MonixSampleClient.scala [hunk 8]
```diff
@@ -17,7 +18,8 @@ object MonifuSampleClient extends js.JSApp {
     val line4 = new DataConsumer(200.millis, -9826395057397L, doBackPressure = false)
       .collect { case s: Signal => s }
 
-    Observable.combineLatest(line1, line2, line3, line4)
+    Observable
+      .combineLatest4(line1, line2, line3, line4)
       .subscribe(new Graph("lineChart"))
```
Comment (alexandru) [inline client/src/main/scala/client/MonixSampleClient.scala:23]: NOTE: this should NOT be `unsafeSubscribeFn`, because here is the call-site (this is the user).

File: client/src/main/scala/client/SimpleWebSocketClient.scala [hunk 9]
```diff
@@ -1,40 +1,40 @@
...
-final class SimpleWebSocketClient private
-  (url: String, os: OverflowStrategy.Synchronous)
+final class SimpleWebSocketClient private(url: String, os: OverflowStrategy.Synchronous[String])
   extends Observable[String] { self =>
 
-  private def createChannel(webSocket: WebSocket)(implicit s: Scheduler) = {
+  private def createChannel(webSocket: WebSocket)(implicit s: Scheduler): Observable[String] = {
     try {
-      val channel = PublishChannel[String](os)
+      val channel = Pipe.publish[String].concurrent(os)
```
Comment (alexandru) [inline client/src/main/scala/client/SimpleWebSocketClient.scala]: Change this to:

``` scala
val (in, out) = Pipe.publish[String].concurrent(os)
```

That's how it was meant to be used :-)

File: client/src/main/scala/client/SimpleWebSocketClient.scala [hunk 10]
```diff
@@ -1,40 +1,40 @@
...
 import scala.concurrent.duration._
 
 
-final class SimpleWebSocketClient private
-  (url: String, os: OverflowStrategy.Synchronous)
+final class SimpleWebSocketClient private(url: String, os: OverflowStrategy.Synchronous[String])
   extends Observable[String] { self =>
 
-  private def createChannel(webSocket: WebSocket)(implicit s: Scheduler) = {
+  private def createChannel(webSocket: WebSocket)(implicit s: Scheduler): Observable[String] = {
```
Comment (alexandru) [inline client/src/main/scala/client/SimpleWebSocketClient.scala]: Everything I said related to the back-pressured observable applies here as well.

Comment (alexandru) [inline client/src/main/scala/client/SimpleWebSocketClient.scala]: Or in other words ... 
1. this `createChannel` should not take a `WebSocket` reference
2. it should create a `new Observable[String]` that internally creates that `WebSocket` reference and the `Pipe.publish` to go along with it
3. etc.

File: client/src/main/scala/client/SimpleWebSocketClient.scala [hunk 11]
```diff
@@ -69,21 +69,21 @@ final class SimpleWebSocketClient private
         scheduler.reportFailure(ex)
         // retry connection in a couple of secs
         self.delaySubscription(3.seconds)
-          .onSubscribe(subscriber)
+          .subscribe(subscriber)
```
Comment (alexandru) [inline client/src/main/scala/client/SimpleWebSocketClient.scala]: `unsafeSubscribeFn`

File: client/src/main/scala/client/SimpleWebSocketClient.scala [hunk 12]
```diff
@@ -69,21 +69,21 @@ final class SimpleWebSocketClient private
...
       }
 
       def onComplete(): Unit = {
         closeConnection(webSocket)
         // retry connection in a couple of secs
         self.delaySubscription(3.seconds)
-          .onSubscribe(subscriber)
+          .subscribe(subscriber)
```
Comment (alexandru) [inline client/src/main/scala/client/SimpleWebSocketClient.scala]: `unsafeSubscribeFn`

File: project/plugins.sbt [hunk 13]
```diff
@@ -10,7 +10,7 @@ resolvers += Resolver.url("heroku-sbt-plugin-releases",
 // Sbt plugins
 addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.4.2")
 
-addSbtPlugin("org.scala-js" % "sbt-scalajs" % "0.6.5")
+addSbtPlugin("org.scala-js" % "sbt-scalajs" % "0.6.8")
```
Comment (alexandru) [inline project/plugins.sbt]: We should upgrade to Scala.js `0.6.9` - it's the latest one, released a week ago or something.

File: server/app/engine/DataProducer.scala [hunk 14]
```diff
@@ -1,33 +1,33 @@
...
+  private case class State(x: Int, y: Int, ts: Long)
 
-  private val points = Observable.create[Signal] { subscriber =>
+  override def unsafeSubscribeFn(subscriber: Subscriber[Signal]): Cancelable = {
     import subscriber.{scheduler => s}
 
     val random = Observable
       .fromStateAction(Random.intInRange(-20, 20))(s.currentTimeMillis() + seed)
-      .flatMap { x => delayedResult(interval)(x) }
+      .flatMap { x => Observable.fromFuture(FutureUtils.delayedResult(interval)(x)) }
```
Comment (alexandru) [inline server/app/engine/DataProducer.scala]: We don't need to go through `Future` here.

``` scala
Observable.now(x).delaySubscription(interval)
```

File: client/src/main/scala/client/SimpleWebSocketClient.scala (hunk 10 above)
Comment (mihaisoloi) [inline client/src/main/scala/client/SimpleWebSocketClient.scala]: OK, but in the case of a SimpleWebSocketClient, I don't understand how exactly we should close the connection when we are returning the `pipe.out`. The only way I have it figured out is by using the `doOnDownstreamStop` method

**Comment by mihaisoloi** [comment]
I think I have fixed your comments, can you take another look-see?

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 15]
```diff
@@ -1,93 +1,103 @@
...
 
-      def onError(ex: Throwable): Unit = {
-        closeConnection(webSocket)
-        scheduler.reportFailure(ex)
-        // retry connection in a couple of secs
-        self.delaySubscription(3.seconds)
-          .onSubscribe(subscriber)
-      }
+        def onComplete(): Unit = {
+          subscription.cancel()
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: This is a "forward reference". The compiler should have already warned you. It's incorrect. The point is that you can end up with a `subscription.cancel()` before the `.subscribe()` above returns. If you don't believe me, test this code with an `Observable.now(1)` which completes immediately. You'll probably get a null pointer exception.
1. Forward references to `Cancelable` objects are created by means of `SingleAssignmentCancelable` - that's its purpose
2. You don't need to cancel the subscription in `onComplete` unless you're doing something terribly wrong, because the source (e.g. `createChannel()`) should already do any required resource cleaning for you ;-)

In other words, delete this `subscription.cancel()`.

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 16]
```diff
@@ -1,93 +1,103 @@
...
-    source.onSubscribe(new Observer[String] {
-      def onNext(elem: String): Future[Ack] =
-        subscriber.onNext(elem)
+    val subscription: Cancelable =
+      createChannel().subscribe(new Observer[String] {
+        def onNext(elem: String): Future[Ack] =
+          subscriber.onNext(elem)
+
+        def onError(ex: Throwable): Unit = {
+          subscription.cancel()
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: Delete this `subscription.cancel()`. Same rationale as below.

File: client/src/main/scala/client/BackPressuredWebSocketClient.scala [hunk 17]
```diff
@@ -1,93 +1,103 @@
...
-    }
-
-    val source = channel.timeout(5.seconds)
-      .doOnCanceled(closeConnection(webSocket))
-
-    source.onSubscribe(new Observer[String] {
-      def onNext(elem: String): Future[Ack] =
-        subscriber.onNext(elem)
+    val subscription: Cancelable =
+      createChannel().subscribe(new Observer[String] {
```
Comment (alexandru) [inline client/src/main/scala/client/BackPressuredWebSocketClient.scala]: This should be an `unsafeSubscribeFn`. Because (1) we know what we are doing and our stuff is protected against contract violations, (2) the normal `subscribe()` wraps subscribers into a `SafeSubscriber` and that's unnecessary when building a new `Observable` (e.g. a new `unsafeSubscribeFn`).

**Comment by alexandru** [comment]
Thanks @mihaisoloi - am merging it!
